Improve support for prioritized HTTPS navigations

sysrqb opened the pull request:
#### b341e0e6cedf800fa1e1a8afed4544cdd0000c27
<pre>
Improve support for prioritized HTTPS navigations
<a href="https://bugs.webkit.org/show_bug.cgi?id=248968">https://bugs.webkit.org/show_bug.cgi?id=248968</a>
rdar://problem/103140804

Reviewed by Alex Christensen.

In this patch we improve the HTTPSOnly behavior that was introduced in an
earlier change. This patch makes three changes:

1. Upgrade main frame navigations except when the navigation is same-site and
   HTTPSOnly is was disabled on this site
2. Upgrade navigation redirects and prevent trivial downgrade redirect loops
3. Reduces the timeout on the data task on cocoa platforms

We also introduce a new HTTPSUpgradeRedirectLoop error to support handling
upgrade-downgrade loops.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::httpsUpgradeRedirectLoopError const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::upgradeRequestforHTTPSOnlyIfNeeded const):
(WebCore::FrameLoader::changeLocation):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
(WebCore::ResourceLoader::httpsUpgradeRedirectLoopError):
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::upgradeInsecureRequestIfNeeded const):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Shared/API/APIError.h:
* Source/WebKit/Shared/WebErrors.cpp:
(WebKit::httpsUpgradeRedirectLoopError):
* Source/WebKit/Shared/WebErrors.h:
* Source/WebKit/UIProcess/API/Cocoa/WKErrorPrivate.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::httpsUpgradeRedirectLoopError const):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::httpsUpgradeRedirectLoopError const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/261022@main">https://commits.webkit.org/261022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c9d18cc3e82c99eefb3e8620de51efd6cdf4cda

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110283 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42927 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1689 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119251 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114232 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20837 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10563 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102576 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116027 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98716 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30380 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12060 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31717 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8684 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18028 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51334 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7636 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14485 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -433,6 +433,18 @@ void FrameLoader::checkContentPolicy(const ResourceResponse& response, PolicyChe
+        FRAMELOADER_RELEASE_LOG(ResourceLoading, "upgradeRequestforHTTPSOnlyIfNeeded: upgrading navigation request");
+        newURL.setProtocol("https"_s);
+        if (newURL.port() && newURL.port().value() == 80)
+            newURL.setPort(443);
```
if (newURL.port() == 80)
    newURL.setPort(std::nullopt);

Same in ContentSecurityPolicy::upgradeInsecureRequestIfNeeded

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -433,6 +433,18 @@ void FrameLoader::checkContentPolicy(const ResourceResponse& response, PolicyChe
 
+bool FrameLoader::upgradeRequestforHTTPSOnlyIfNeeded(const URL& originalURL, URL& newURL) const
+{
+    if (m_documentLoader && m_documentLoader->networkConnectionIntegrityPolicy().contains(NetworkConnectionIntegrity::HTTPSOnly) && newURL.protocolIs("http"_s) && ((RegistrableDomain(newURL) == RegistrableDomain(originalURL) && !m_documentLoader->networkConnectionIntegrityPolicy().contains(NetworkConnectionIntegrity::HTTPSOnlyExplicitlyBypassedForDomain)) || RegistrableDomain(newURL) != Registrab ...
```
split up

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -457,6 +469,11 @@ void FrameLoader::changeLocation(FrameLoadRequest&& frameRequest, Event* trigger
         frameRequest.setFrameName(m_frame.document()->baseTarget());
 
+    auto currentURL { m_frame.document() ? m_frame.document()->url() : URL { } };
+    auto url { frameRequest.resourceRequest().url() };
```
newURL?

achristensen07 commented on Source/WebCore/loader/ResourceLoader.cpp:
```diff
@@ -410,6 +410,19 @@ void ResourceLoader::willSendRequestInternal(ResourceRequest&& request, const Re
         return;
     }
 
+    if (auto isMainFrame = (frameLoader() ? frameLoader()->frame().isMainFrame() : false); isMainFrame && cachedResource() && cachedResource()->type() == CachedResource::Type::MainResource && !redirectResponse.isNull()) {
```
isMainFrame not needed

achristensen07 commented on Source/WebCore/loader/ResourceLoader.cpp:
```diff
@@ -410,6 +410,19 @@ void ResourceLoader::willSendRequestInternal(ResourceRequest&& request, const Re
     }
 
+    if (auto isMainFrame = (frameLoader() ? frameLoader()->frame().isMainFrame() : false); isMainFrame && cachedResource() && cachedResource()->type() == CachedResource::Type::MainResource && !redirectResponse.isNull()) {
+        auto requestURL { m_request.url() };
```
redirectResponse.url()
Less confusing.  request and m_request

achristensen07 commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -415,6 +415,10 @@ static inline bool computeIsAlwaysOnLoggingAllowed(NetworkSession& session)
 
+    // FIXME: Make this timeout adaptive based on network conditions
+    if (url.protocolIs("https"_s) && networkConnectionIntegrityPolicy.contains(WebCore::NetworkConnectionIntegrity::HTTPSOnly))
+        mutableRequest.get().timeoutInterval = 3.0f;
```
.0f unneeded

sysrqb committed:
Improve support for prioritized HTTPS navigations
https://bugs.webkit.org/show_bug.cgi?id=248968
rdar://problem/103140804

Reviewed by Alex Christensen.

In this patch we improve the HTTPSOnly behavior that was introduced in an
earlier change. This patch makes three changes:

1. Upgrade main frame navigations except when the navigation is same-site and
   HTTPSOnly is was disabled on this site
2. Upgrade navigation redirects and prevent trivial downgrade redirect loops
3. Reduces the timeout on the data task on cocoa platforms

We also introduce a new HTTPSUpgradeRedirectLoop error to support handling
upgrade-downgrade loops.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::httpsUpgradeRedirectLoopError const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::upgradeRequestforHTTPSOnlyIfNeeded const):
(WebCore::FrameLoader::changeLocation):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
(WebCore::ResourceLoader::httpsUpgradeRedirectLoopError):
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::upgradeInsecureRequestIfNeeded const):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Shared/API/APIError.h:
* Source/WebKit/Shared/WebErrors.cpp:
(WebKit::httpsUpgradeRedirectLoopError):
* Source/WebKit/Shared/WebErrors.h:
* Source/WebKit/UIProcess/API/Cocoa/WKErrorPrivate.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::httpsUpgradeRedirectLoopError const):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::httpsUpgradeRedirectLoopError const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: https://commits.webkit.org/261022@main
